File: julia/collatz-conjecture/collatz-conjecture.jl
function collatz_steps(n)
    n < 1 && throw(DomainError("Number must be positive"))

    steps = 0

    while n != 1
        n = iseven(n) ? n ÷ 2 : n * 3 + 1
        steps += 1
    end

    steps
end

File: prolog/collatz-conjecture/collatz_conjecture.pl
collatz_steps(1, 0) :- !.

collatz_steps(N, Steps) :- 
    N > 1,
    (0 is mod(N, 2), M is (N / 2); M is (N * 3 + 1)),
    collatz_steps(M, MSteps),
    Steps is MSteps + 1,
    !.
